### DeepVision Classifier

In [1]:
import torch 
import torch.nn as nn
import torch.optim as optim

import torchvision
from torchvision.datasets import CIFAR10

In [2]:
#Datasets and DataLoaders
from torch.utils.data import DataLoader
import torchvision.transforms as transforms


#image -> scale(0,1) -> normalize(-1,1).....This helps the data to be centered more around zero
transform=transforms.Compose([         
    transforms.ToTensor(),
    transforms.Normalize((0.5,0.5,0.5),(0.5,0.5,0.5))
])    

trainset=CIFAR10(root="./data",train=True,download=False,transform=transform)
testset=CIFAR10(root="./data",train=False,download=False,transform=transform)


#Compose = “Do these steps one after another in order

In [3]:
trainLoader=DataLoader(trainset,batch_size=64,shuffle=True)
testLoader=DataLoader(testset,batch_size=64,shuffle=True)


### CNN MODULE

In [26]:
class CNN(nn.Module):
    def __init__(self):
        super(CNN,self).__init__()


        self.conv_layers = nn.Sequential(
            nn.Conv2d(3,32,kernel_size=3,padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2,2),#(kernal_size,stride)

       
            nn.Conv2d(32,64,kernel_size=3,padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2,2),#(kernal_size,stride)
            
       
            nn.Conv2d(64,128,kernel_size=3,padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2,2)#(kernal_size,stride)
            
        )

        self.fc_layers=nn.Sequential(
            nn.Linear(4*4*128,256),
            nn.ReLU(),
            nn.Linear(256,10)
        )
        
    def forward(self,x):
      x=self.conv_layers(x)
      x=x.view(x.size(0),-1)
      x=self.fc_layers(x)
      return x   
        


In [27]:
model=CNN()
criterion=nn.CrossEntropyLoss()
optimizer=optim.Adam(model.parameters())


In [30]:
epochs=10

for epoch in range(epochs):
    epoch_training_loss=0.0

    for images,labels in trainLoader:
        optimizer.zero_grad()

        output=model.forward(images)
        loss=criterion(output,labels)
        loss.backward()
        optimizer.step()

        epoch_training_loss+=loss.item()

    print(f"epoch={epoch+1} & loss={epoch_training_loss/len(trainLoader)}")

epoch=1 & loss=0.1260512186871732
epoch=2 & loss=0.11877205847498135
epoch=3 & loss=0.09297780234299963
epoch=4 & loss=0.08357148743746683
epoch=5 & loss=0.09270561541921324
epoch=6 & loss=0.07504206826991361
epoch=7 & loss=0.08314417652509477
epoch=8 & loss=0.06589412638876478
epoch=9 & loss=0.06790354894861207
epoch=10 & loss=0.07000826482135264


In [31]:
correct_labels = 0
total_labels = 0

with torch.no_grad():
    for images, labels in testLoader:
        outputs = model(images)
        _, predicted = torch.max(outputs, dim=1)
        correct_labels += (predicted == labels).sum().item()
        total_labels += labels.size(0)

accuracy = (correct_labels / total_labels) * 100

print(f"Test Accuracy: {accuracy:.2f}%")

Test Accuracy: 74.43%


->Built a Convolutional Neural Network (CNN) for multi-class image classification using the CIFAR-10 dataset.
Applied preprocessing techniques like normalization to improve training performance.
    
-> Designed a model with multiple convolutional and fully connected layers for feature extraction and classification.
Evaluated the model and achieved an accuracy of 74.43%.

